docs/yum-torrent.1
.TH yum-torrent 1 "09 MARCH 2016"
.SH NAME
yum-torrent \- download packages via http and peer-to-peer
.SH DESCRIPTION
.B yum-torrent
extends the behaviour of yum by downloading then seeding packages via BitTorrent.
Once installed and configured, the plugin works transparently.
.SH FILES
.B yum-torrent
makes extensive use of cache files. The BitTorrent client's cwd is set to the root of the cache by default.
All generated .torrent files are guaranteed to contain files in subdirectories.
Once a torrent download is complete the files are hard or soft linked to the path in
.B
/var/cache/yum
where yum will look before starting it's own downloads, bypassing it.
.PP
.B yum-torrent
will automatically clean up unused files and directories left over from previous runs.
.PP
.RS
.nf
/var/cache/yum_torrent/\fItorrenthash\fR.torrent
/var/cache/yum_torrent/\fItorrenthash\fR.torrent.log
/var/cache/yum_torrent/\fIpath/to/package\fR.rpm
/var/cache/yum/\fIrepo\fR/packages/\fIpackage\fR.rpm
.fi
.RE
.SH AUTHOR
Matthew Almond <malmond@fb.com>
.SH SEE ALSO
yum-torrent.conf(5),
pkgtorrent-service(8),
pkgtorrent-service.conf(5)


ChangeLog
2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* ChangeLog: updated ChangeLog

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-03  Shawn Starr <shawn.starr@rogers.com>

	* docs/yum-utils.1: Update yum-utils man page, to add new utils.

2011-07-25  James Antill <james@and.org>

	Merge branch 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils  * 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils: (3 commits)
	typo leaving .conf in ybc.conf.yumvar should just be ybc.yumvar
	...

2011-07-25  James Antill <james@and.org>

	* plugins/aliases/aliases: Add some new aliases, fix chkD.

2011-07-25  Seth Vidal <skvidal@fedoraproject.org>

	* yum-config-manager.py: typo leaving .conf in ybc.conf.yumvar
	should just be ybc.yumvar

2011-07-12  James Antill <james@and.org>

	* plugins/security/security.py: Also auto-apply --security when
	given --sec-severity for other commands.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add repoquery @group completion.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* repoquery.py: Fix "nitmes" typo in previous commit.

2011-07-06  James Antill <james@and.org>

	* repoquery.py: Fix typo

2011-06-29  James Antill <james@and.org>

	* repoquery.py: Allow @foo groups as a packages selector in
	repoquery.

2011-06-25  Ville Skyttä <ville.skytta@iki.fi>

	* plugins/aliases/aliases.py, plugins/security/security.py,
	plugins/versionlock/versionlock.py: Remove some unused local
	variables and imports flagged by pyflakes.

2011-06-24  Ville Skyttä <ville.skytta@iki.fi>

	* debuginfo-install.py, package-cleanup.py, yum-builddep.py,
	yum-complete-transaction.py: Check doUtilBuildTransaction return
	value, exit early on failure.  BZ 716267

2011-06-22  James Antill <james@and.org>

	* plugins/security/security.py: Try to be clever when sorting the
	security severities.

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: After (external) merge is
	complete, give user option to delete other file

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: Support meld as well as vimdiff
	for merging  When using meld, the left side has the "other file"
	i.e. the file with the .rpmnew or .rpmsave extension. The right side
	has the "final file" i.e. the target file into which changes have to
	be merged.  If the rpm action was to save the existing conf file,
	then the left side will be the original conf file (.rpmsave) and the
	right side will be the new rpm-based configuration. The user should
	merge any changes they want to keep from their old .rpmsave from
	left to right.  If the rpm action was to keep the existing conf
	file, then the left side will be the new conf file (.rpmnew) and
	right side will be the current local configuration. The user should
	merge any changes in the new rpm file into their local config from
	left to right.  Ideally these operations would be 3-way diffs
	between the local file and the new rpm config, using the original
	rpm config as the common merge base. However, the original rpm
	config is not available to yum and therefore a 3-way diff is not
	possible. Perhaps a future enhancement would be to use a .orig file,
	if present, as the merge base for meld.  When using vimdiff, the
	order has not changed. The first parameter to vimdiff is the "final:
	file i.e. the target file, and the second parameter is the "other"
	file i.e. the .rpmnew or .rpmsave file.

2011-06-10  James Antill <james@and.org>

	* repodiff.py: Huge speedup for large number of add/remove obsoletes
	processing.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* yumdownloader.py: Stops yumdownloader from attempting to download
	the same package twice if it is available in multiple repos. BZ
	711767

2011-06-10  James Antill <james@and.org>

	* docs/repodiff.1, repodiff.py: Add --compare-arch for the other
	side of the multilib. diff. BZ 710579.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* repodiff.py: Changed archlist set up to handle multilib. BZ 710579


2011-06-01  James Antill <james@and.org>

	* docs/repoquery.1: Add releasever and setopt docs. from yum to
	repoquery. BZ 638542.

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* yum-utils.spec: frelling pyyaml pkg named with weird case

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/puppetverify/puppetverify.conf,
	plugins/puppetverify/puppetverify.py, yum-utils.spec: add
	puppetverify plugin

2011-05-18  Frank Weltersbach <frankwe@live.com>

	* docs/package-cleanup.1: Fix typo in package-cleanup man page. BZ
	705617.

2011-05-17  Seth Vidal <skvidal@fedoraproject.org>

	* repoquery.py:     - fix rh bug: #705179     - fix fmt_queryformat not being happy
	    with new 'output' formats passed       everywhere else.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Allow files/provides to be used in
	yumdownloader.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Exit with code 2, if we can't download the
	packages asked for. BZ 699469.

2011-04-26  James Antill <james@and.org>

	* plugins/auto-update-debuginfo/auto-update-debuginfo.py: try/except
	the open, for the rpmdbv cached data (a-u-d plugin).

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add yum-debug-dump completion.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Mention filename arg in yum-debug-dump usage
	message.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Make yum-debug-dump.py executable for running
	in-place.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Use functions from recent bash-completion if
	available.

2011-04-06  James Antill <james@and.org>

	* repoquery.py, yum-groups-manager.py: Convert

plugins/pkgtorrent/client/yum-torrent.py
#
# Copyright (c) 2014-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the GPLv2 license found in the COPYING
# file in the root directory of this source tree.
#
from urlparse import urlsplit, urlunsplit
from yum.plugins import TYPE_CORE
from collections import defaultdict
from hashlib import sha1
import subprocess
import httplib
import urllib
import urllib2
import struct
import socket
import stat
import time
import traceback
import os

requires_api_version = '2.1'
plugin_type = (TYPE_CORE,)
CONTENT_DISPOSITION_PREFIX = 'attachment; filename='
PROC = '/proc'
# a format for identifying pre-existing torrent processses using argv[0]
DOWNLOAD_PROCESS_FORMAT = 'yum_torrent_downloading_%s'
# the directory managed by this program. At one point I thought of making
# it configurable, but because this program can do cache cleanups of files
# it's too dangerous to expose by default.
TORRENTBASE = '/var/cache/yum_torrent'
service, client, options, complete = None, None, None, None
timeout_download, timeout_nullfile = None, None
timeout_seed_min, timeout_seed_max = None, None
timeout_cache = None

def show_waiting(conduit, torrentlogs):
    n = len(torrentlogs)
    if n == 0:
        conduit.info(2, 'Download complete.')
        return
    if n > 1:
        s = 's'
    else:
        s = ''
    conduit.info(2, 'Waiting on %d torrent%s.' % (n, s))

def init_hook(conduit):
    global service, client, options, complete
    global timeout_download, timeout_nullfile
    global timeout_seed_min, timeout_seed_max
    global timeout_cache

    service = conduit.confString(
        'main',
        'service',
        default='/yum/torrent_service',
    )
    client = conduit.confString('main', 'client')
    options = conduit.confString('main', 'options', default='')
    complete = conduit.confString(
        'main',
        'complete',
        default='Download complete',
    )
    timeout_download = conduit.confInt('timeouts', 'download', default=200)
    timeout_seed_min = conduit.confInt('timeouts', 'seed_min', default=900)
    timeout_seed_max = conduit.confInt('timeouts', 'seed_max', default=86400)
    timeout_nullfile = conduit.confInt('timeouts', 'nullfile', default=86400)
    timeout_cache = conduit.confInt('timeouts', 'cache', default=172800)
    if timeout_seed_max < timeout_seed_min:
        conduit.info(
            2,
            'Warning: Seed maximum time (%d) is less than the minimum (%d). '
            'Forcing the maximum to match the minimum.' % (
                timeout_seed_max,
                timeout_seed_min,
            )
        )
        timeout_seed_max = timeout_seed_min
    if timeout_cache < timeout_seed_max:
        conduit.info(
            2,
            'Warning: Cache directory cleanup timeout (%d) is less than the '
            'maximum seed time (%d). Setting the cache timeout to the maximum '
            'seed time so we do not delete files while we could be seeding '
            'them.' % (timeout_cache, timeout_seed_max),
        )
        timeout_cache = timeout_seed_max

def catchall(f):
    def inner(conduit):
        try:
            return f(conduit)
        except Exception as e:
            conduit.info(
                    2,
                    'Unhandled exception: %s Trace: %s' % (str(e),
                        traceback.format_exc()))
    return inner

@catchall
def predownload_hook(conduit):
    hostpaths = defaultdict(list)
    for package in conduit.getDownloadPackages():
        try:
            result = urlsplit(package.remote_url)
            if result.scheme not in ('http', 'https'):
                # whitelist http urls because we can POST to them. This will
                # disable the plugin for ftp and file protocols.
                continue
            # normal yum plugin passed objects with this member; note that
            # yum/sqlitesack.py overrides __getattr__ for this object with an
            # implementation that doesn't support default fallbacks, and throws
            # a KeyError exception when an invalid attribute is accessed; this
            # is why we have to do this dance.
            if hasattr(package, 'localpath'):
                localpath = getattr(package, 'localpath')
            else:
                # if not present, then we're running from Anaconda, which uses
                # yum.sqlitesack.YumAvailablePackageSqlite instances.
                localpath = package.localPkg()
            hostpaths[result.scheme, result.netloc].append((
                result.path,
                package.size,
                localpath,
            ))
        except ValueError:
            # urlsplit() will choke on localinstall packages. There isn't
            # a more elegant solution. In any case, if this code breaks, the
            # fallback is to download normally which is fine.
            pass

    # lame enumeration of process argv[0]. We used this later to work out if
    # we're already downloading said torrent.
    process_names = {}
    for pid in os.listdir(PROC):
        try:
            # We want to limit to only numeric directories, and we'll store
            # the number, not string later.
            process_names[
                open(os.path.join(PROC, pid, 'cmdline')).read().split('\0')[0]
            ] = pid
        except:
            # hey, we tried.
            pass

    torrentlogs = {}
    for (scheme, netloc), packages in hostpaths.iteritems():
        packages.sort()
        packagepaths = [p[0] for p in packages]
        digest = sha1('\n'.join(packagepaths)).hexdigest()
        url = urlunsplit(
            (scheme, netloc, service + '/digest/' + digest, '', '')
        )
        try:
            conduit.info(2, 'GET URL: %s' % (url,))
            res = urllib2.urlopen(url)
        except urllib2.HTTPError as e:
            if e.getcode() == 412:
                conduit.info(
                    2,
                    "Service doesn't know about this digest, POST'ing paths..."
                )
                data = urllib.urlencode([('p', p) for p in packagepaths])
                try:
                    res = urllib2.urlopen(url, data)
                except urllib2.HTTPError as e:
                    conduit.info(
                        2,
                        'Service failed %d, skipping torrent' % (e.getcode(),)
                    )
                    continue
                except (httplib.HTTPException, urllib2.URLError) as e:
                    conduit.info(
                        2,
                        'Service failed: %s, skipping torrent' % (str(e)))
                    continue
            else:
                conduit.info(
                    2,
                    "Server returned %d. Perhaps service isn't installed?" % (
                        e.getcode(),
                    ),
                )
                continue

        code = res.getcode()
        if code != 200:
            conduit.info(2, 'Torrent service returned %s' % (code,))
            continue
        h = res.info()
        cd = res.info()['Content-Disposition']
        assert cd.startswith(CONTENT_DISPOSITION_PREFIX)
        filename = os.path.basename(cd[len(CONTENT_DISPOSITION_PREFIX):])
        open(os.path.join(TORRENTBASE, filename), 'w').write(res.read())

        # work out the seed time. It's 1/x shaped and biased by
        # timeout_seed_min and timeout_seed_max, and is salted by the digest
        # of the files specified, so different sets of files will be seeded
        # by different hosts.
        f = sha1(socket.gethostname())
        f.update(digest)
        # this requires some explanation: The digest for a sha1 returns
        # 20 bytes. I want to turn this into a float between 0 and 1. To
        # do this, I take the first 4 bytes, convert them into an "standard"
        # (4 byte) unsigned int, then divide by the max unsigned int.
        shard = struct.unpack('=L', f.digest()[:4])[0] / 2.0 ** 32
        # the seed time is a curve from 0% of hosts doing the max timeout
        # to 100% of hosts doing the minimum timeout. See
        # https://fburl.com/yumtorrentcurve
        # for the shape. Instead of importing the math library, I'm raising to
        # the 0.5th power, which is the same thing.
        seed = timeout_seed_min + (timeout_seed_max - timeout_seed_min) * \
                (1 - (-(shard - 2) * shard) ** 0.5)

        # replace '{seed_time}' in options (if present) with the number of
        # seconds to seed for.
        args = options.replace('{seed_time}', str(int(seed))).split()
        log = os.path.join(TORRENTBASE, filename + '.log')
        process_name = DOWNLOAD_PROCESS_FORMAT % (filename,)
        if process_name in process_names and os.path.isfile(log):
            pid = process_names[process_name]
            conduit.info(
                2,
                'PID %s is downloading/seeding %s, not spawning new client.' % (
                    pid,
                    filename,
                )
            )
        else:
            subprocess.Popen(
                [process_name] + args + [filename],
                executable=client,
                stdout=open(log, 'w'),
                cwd=TORRENTBASE,
                close_fds=True,
            )

        # let's say all the files were last modified at the EPOCH which isn't
        # true but this means each will get completely read at startup.
        torrentlogs[open(log)] = ''

    download_until = time.time() + timeout_download
    # it's possible you end up with the complete message across loops. Keep
    # up to this much data from previous runs.
    extra_data_len = len(complete) - 1
    show_waiting(conduit, torrentlogs)
    while torrentlogs:
        # note we're modifying the torrentlogs dict but this is a copy.
        # this will need to be changed for python3
        for log, olddata in torrentlogs.items():
            # a single read() may not return everything so keep doing it
            # until you get nothing.
            data = [olddata]
            while True:
                # read alllllll the data
                moar = log.read()
                if moar == '':
                    # eof, that'll do.
                    break
                data.append(moar)
            data = ''.join(data)
            if data.find(complete) != -1:
                log.close()
                del torrentlogs[log]
                show_waiting(conduit, torrentlogs)
                continue
            # update the tail of data last seen.
            torrentlogs[log] = data[-extra_data_len:]

        if time.time() > download_until:
            conduit.info(
                2,
                'Timed out waiting for bittorent download completion. '
                'Falling back on normal download semantics.',
            )
            # let's just bail on everything. We can't reliably do any
            # symlinking or nullfile cleanup. Note we keep trying to
            # download into the yum_torrent directory and seed as normal.
            return
        # we can't use poll() so we'll pause for a moment for things to change.
        time.sleep(1)

    # note client can run longer than this, and should, so it keeps seeding.

    nulldir = os.path.join(TORRENTBASE, service[1:], 'nullfile')
    if os.path.isdir(nulldir):
        age_threshold = time.time() - timeout_nullfile
        for filename in os.listdir(nulldir):
            nullfile = os.path.join(nulldir, filename)
            s = os.stat(nullfile)
            if s.st_mtime < age_threshold:
                conduit.info(3, 'Deleting nullfile: %s' % (nullfile,))
                os.unlink(nullfile)
                continue
            # rough heuristic for spotting sparse files.
            if s.st_blocks > (s.st_blksize / 512):
                tmpnullfile = '%s.sparse' % (nullfile,)
                f = open(tmpnullfile, 'wb')
                f.seek(s.st_size - 1)
                f.write('\0')
                f.close()
                # atomically replace
                try:
                    os.rename(tmpnullfile, nullfile)
                except OSError:
                    os.unlink(tmpnullfile)

    cutoff = time.time() - timeout_cache
    for packages in hostpaths.itervalues():
        for name, size, localpath in packages:
            downloadpath = os.path.join(TORRENTBASE, name[1:])
            try:
                if os.stat(downloadpath).st_mtime < cutoff:
                    # This file was about to expire, but let's keep it
                    # alive as we just referenced it again.
                    os.utime(downloadpath, None)
            except OSError as e:
                if e.errno == 2:
                    # file not found. We can't touch it, and shouldn't link
                    # to it either.
                    conduit.info(
                        2,
                        'Warning: Download file not found: %s' % (downloadpath,)
                    )
                    continue
                raise

            if os.path.exists(localpath):
                continue
            localdir = os.path.dirname(localpath)
            if not os.path.isdir(localdir):
                os.makedirs(localdir)
            try:
                os.link(downloadpath, localpath)
            except OSError as e:
                if e.errno != 18:
                    # not a cross device hardlink error, can't handle this.
                    raise
                os.symlink(
                    os.path.relpath(downloadpath, localdir),
                    localpath,
                )

@catchall
def clean_hook(conduit):
    # Let's go look at all of TORRENTBASE and clean up files older than
    # timeout_cache and remove empty directories.
    cutoff = time.time() - timeout_cache
    for root, dirs, files in os.walk(TORRENTBASE, topdown=False):
        for name in files:
            fullpath = os.path.join(root, name)
            if os.stat(fullpath).st_mtime >= cutoff:
                continue
            # note this doesn't try to work out what is symlinked to this file.
            # On the unlikely event that we end up breaking a symlink, then
            # yum will download a file and replace the symlink with a real file.
            conduit.info(2, 'Deleting expired cache file: %s' % (fullpath,))
            os.unlink(fullpath)

            # TODO: also delete empty directories.
        for name in dirs:
            fullpath = os.path.join(root, name)
            if os.listdir(fullpath):
                continue
            conduit.info(2, 'Deleting empty directory: %s' % (fullpath,))
            os.rmdir(fullpath)
